### ISSUE TITLE
Add logout button in web apps

### ISSUE THREAD

**Issue by pmarini** [mjl-/mox#58]
I noticed that in none of the web apps (`admin`, `webmail`, `account`) there is a logout button.

It would be useful to have it.

Thanks

**Comment by mjl-** [comment]
Agreed, would be good to have.

Mox currently uses http basic authentication. It is awfully easy to implement and works surprisingly well (though a scheme were the plaintext password wouldn't be sent to the server would be a lot better!). Anyway, I had a quick look, and it looks like there isn't a good way to clear the HTTP auth credentials, see https://stackoverflow.com/questions/233507/how-to-log-out-user-from-web-site-using-basic-authentication. Though it would be good to experiment with the approaches mentioned there.

**Comment by mjl-** [comment]
I tried an approach based on the stackoverflow posts. That is, to logout navigate the user to an endpoint that returns "401", causing the browser to stop the authenticated session and try to reauthenticate. I thought it was going to work. But turned out it won't, because one another open tab keeps the authenticated session alive (at least in Safari). Just now it dawned on me why I haven't even thought about the need for this: I use Firefox with a plugin that opens each new tab in a new session/container (because the web is a hostile place...). So I don't get any authenticated session sharing when I'm testing... This is making me realize that having a way to logout is more important for many people than I thought. The solution will be to do a more typical login page...

**Comment by x8x** [comment]
Probably worth mentioning in the Roadmap that a non-HTTP-basic-authentication login is already planned!

**Comment by pmarini** [comment]
NIce to hear!

**Comment by jsfan3** [comment]
For me, not having a logout button is a big problem because I have to create multiple accounts on mox to use webmail. As it is now, once I am logged into one account, there is no way to switch to another account.

I know that with private browsing I can log out by closing the browser..., but this is only a temporary workaround and only for experienced users.

**Comment by jsfan3** [comment]
Maybe a possible solution without too much effort:
https://stackoverflow.com/a/32325848

**Comment by mjl-** [comment]
The latest commits replaces the http basic auth-based login with cookie/csrf session-based authentication, with a custom login screen. And with a logout button. I think that will solve this issue. Testing is always welcome. You can easily test locally:

```
GOBIN=$PWD CGO_ENABLED=0 go install github.com/mjl-/mox@c348834ce94a35c49a121034d89169139602b5b3
./mox localserve
```

**Comment by jsfan3** [comment]
This is great news! Thank you for your efforts.

As for testing, I will let you know soon.

The installation instructions are not clear to me. Why mox@c348834ce94a35c49a121034d89169139602b5b3 and not mox@0f8bf2f220887078d4f3b54efff643c93580e9ed ?

**Comment by mjl-** [comment]
c348834ce94a35c49a121034d89169139602b5b3 is the latest commit. there were a few after the one that referenced this issue. in between is at least one commit that is somewhat relevant: https://github.com/mjl-/mox/commit/9796c4539d6aed8e8ba1d8ab183002eba5d65d93

**Comment by jsfan3** [comment]
I tested mox v0.0.9-0.20240105111555-c348834ce94a

The new login/logout is much better and allows me to switch between accounts.

Great job! 

I suggest making the box for entering email and password wider. In my case, I have to enter a 40 character email address and the current form is a bit tight.

**Comment by pmarini** [comment]
Tested on newly released 0.0.9:
- login / logout to /admin from Firefox
- login / logout to /webmail from Firefox
- login / logout to /admin from Chromium
- login / logout to /webmail from Chromium

browser versions are stock in Ubuntu Desktop 22.04.

Everything ok! Thanks a lot

**Comment by mjl-** [comment]
Awesome, thanks for testing jsfan3 & pmarini. Closing this ticket, feel free to reopen if needed.
